### ISSUE TITLE
Help #1 : comment s'en sortir avec les options et les structures...

### ISSUE THREAD

**Issue by bfovet** [bfovet/minimesh#3]
Selon le type que vous allez choisir de retourner depuis `OptionsParser::parse` (c'est-à-dire la struct `Params`, un pointeur sur la struct `Params` ou même un pointeur intelligent...), il vous faudra instancier dans chaque condition le type adéquat.

Par ex., j'ai choisi de retourner un `shared_ptr` de `Params`, dans la condition où la transformation est une fusion de maillage :

    std::shared_ptr<MergeParams> merge_params = std::make_shared<MergeParams>();
    merge_params->transform_type = TransformType::MERGE;

    // TODO: remplir merge_params avec les autres options

    return merge_params

Maintenant, dans le `main`, il va falloir que je récupère soit `MergeParams`, soit `TranslateParams`; or je renvoie de `OptionsParser::parse` un std::shared_ptr<Params>. D'où la conversion avec un `dynamic_pointer_cast`. C'est aussi faisable avec un pointeur classique ou sans pointeur du tout, cherchez "c++ cast base class to derived class" sur internet...

    // Dans la condition où command == "--transform"
    auto op = std::make_shared<OptionsParser>(file_name);
    auto params = op->parse();

    // Si params->transform_type est du type MERGE
    std::shared_ptr<MergeParams> mp = std::dynamic_pointer_cast<MergeParams>(params);

**Comment by qborbeau** [comment]
Dans le cas où l'on souhaite se débrouiller sans pointeur, comment faire pour sortir les paramètres du fichier .toml (avec la fonction _parse_) ? Faut-il modifier cette fonction parse justement ?
L'objectif est ensuite d'utiliser ces paramètres dans une fonction _translate_ et une fonction _merge_ c'est bien ça ?

**Comment by bfovet** [comment]
> Dans le cas où l'on souhaite se débrouiller sans pointeur, comment faire pour sortir les paramètres du fichier .toml (avec la fonction _parse_) ? Faut-il modifier cette fonction parse justement ?

Oui, le premier exemple est un bout de code de `OptionsParser::parse`, il faut que tu la modifies de la même façon, en retournant une struct au lieu d'un pointeur. Mais attention, sans pointeur tu perdras tes données si tu ne définies pas ta struct `Params` en dehors de la fonction `parse` (à cause du scope : https://stackoverflow.com/questions/19205024/how-to-return-a-struct-from-a-function-in-c).
Et un autre lien pour savoir comment caster des struct Parent <-> Child : https://www.bogotobogo.com/cplusplus/upcasting_downcasting.php

> L'objectif est ensuite d'utiliser ces paramètres dans une fonction _translate_ et une fonction _merge_ c'est bien ça ?

Oui ! A partir du 2e bout de code (dans `main`), on a toutes les options dans `MergeParams` (ou `TranslateParams`) à fournir ensuite à `Transform::merge` (ou `Transform::translate`).

**Comment by salouaraqa** [comment]
J'ai toujours du mal avec le shared_ptr, si je fais std::shared_ptr<MergeParams> mp = std::dynamic_pointer_cast<MergeParams>(params); je récupère bien mp (qui est du coup MergeParams) que je peux passer à ma fonction merge ?

**Comment by bfovet** [comment]
> J'ai toujours du mal avec le shared_ptr, si je fais std::shared_ptr mp = std::dynamic_pointer_cast(params); je récupère bien mp (qui est du coup MergeParams) que je peux passer à ma fonction merge ?

Presque, tu récupères un `std::shared_ptr<MergeParams>`. J'ai utilisé un `shared_ptr` dans mon exemple mais on peut très bien s'en sortir avec des pointeurs classiques : la ligne équivalente serait `MergeParams *mp = dynamic_cast<MergeParams*>(params);`

A donner à la méthode merge pour récupérer les paramètres à fournir aux filtres VTK.
